### PR TITLE
Add formatting checks to DLC repo

### PR DESCRIPTION
**Pull request by arjkesh**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
- Standardize formatting across all files (use black formatting from https://github.com/psf/black)
- Run black -l 100 on all python files in repo
- Add pyproject.toml to add default configuration
- Following guidelines from https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html and https://black.readthedocs.io/en/stable/faq.html#is-black-safe-to-use

### Tests run
- Add new quick checks test to ensure that formatting is met (test_formatting_and_linting.py)

ToDo for future - add pylint checks (out of scope for now)

I will follow up with another commit to add a .git-blame-ignore-revs file once this is merged in (see https://github.com/aws/deep-learning-containers/pull/2830/files) to avoid overwriting git blame history

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by Lokiiiiii — DISMISSED** [review]

File: test/dlc_tests/sanity/quick_checks/test_formatting_and_linting.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    Check that black style formatting exits gracefully.
+    """
+    # Look up the path until deep-learning-containers is in our base directory
+    dlc_base_dir = os.getcwd()
+    while "deep-learning-containers" not in os.path.basename(dlc_base_dir):
+        dlc_base_dir = os.path.split(dlc_base_dir)[0]
+
+    ctx = Context()
+
+    ctx.run(f"black --check {dlc_base_dir}", hide=True)
```
Comment (kace) [inline test/dlc_tests/sanity/quick_checks/test_formatting_and_linting.py:27]: Iirc `black` logs errors to stderr. It might be worthwhile to include this output in our test logging by changing the `hide` kwarg value. Reason being that different versions of `black` can sometimes have differences in formatting requirements.

I have personal experience where `black` failed in pipelines and passed locally for this exact reason. It can be very confusing and having any sort of error output from the tool would provide some context for a formatting check failure.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -19,6 +19,9 @@ Note: If merging this PR should also close the associated Issue, please also add
 
 **Additionally, please run the sagemaker local tests in at least one revision:**
 - [ ] `sagemaker_local_tests = true`
+  
+### Formatting
+I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)
```
Comment (kace) [inline .github/PULL_REQUEST_TEMPLATE.md:24]: Is this meant to be a checkbox? If so, it's missing.

File: test/dlc_tests/sanity/quick_checks/test_yml_config.py [hunk 3]
```diff
@@ -34,24 +34,25 @@ def _release_images_yml_verifier(image_type, excluded_image_type):
...
+    while "deep-learning-containers" not in os.path.basename(dlc_base_dir):
         dlc_base_dir = os.path.split(dlc_base_dir)[0]
 
     release_images_yml_file = os.path.join(dlc_base_dir, f"release_images_{image_type}.yml")
 
     # Define exclude regex
     exclude_pattern = re.compile(rf"{excluded_image_type}", re.IGNORECASE)
 
-    with open(release_images_yml_file, "r") as release_imgs_handle:
+    with open(release_images_yml_file, "r", encoding="utf-8") as release_imgs_handle:
```
Comment (kace) [inline test/dlc_tests/sanity/quick_checks/test_yml_config.py:47]: [nit] Why stipulate encoding here? My understanding is that `utf-8` is the Python default.

Comment (kace) [inline test/dlc_tests/sanity/quick_checks/test_yml_config.py:47]: [nit] Why stipulate encoding here? My understanding is that `utf-8` is the Python default.

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,2 @@
+[tool.black]
+line-length = 100
```
[the remark this answers is not in the document]
Comment (arjkesh) [reply pyproject.toml:2]: pylint default is 100, and PEP8 line length is 79 chars. I went with the longer line default because we have typically been using 120 chars

File: test/dlc_tests/sanity/quick_checks/test_formatting_and_linting.py (hunk 1 above)
Comment (arjkesh) [reply test/dlc_tests/sanity/quick_checks/test_formatting_and_linting.py:27]: hide=True is necessary to keep output with pytest stacktrace when running tests in parallel. It is an edge case with invoke/pytest behavior, you'll see that throughout our codebase.

It is not as valuable, imo to have the output print somewhere high up in the logs as opposed to along with the stack trace. So as implemented, I think it it will be easier to work with.

File: test/dlc_tests/sanity/quick_checks/test_yml_config.py (hunk 3 above)
Comment (arjkesh) [reply test/dlc_tests/sanity/quick_checks/test_yml_config.py:47]: I believe encoding is actually defaulted to system default. See https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/unspecified-encoding.html - 

"It is better to specify an encoding when opening documents. Using the system default implicitly can create problems on other operating systems. See https://peps.python.org/pep-0597/"

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (arjkesh) [reply .github/PULL_REQUEST_TEMPLATE.md]: It prob makes more sense as a checkbox. Added.

**Review by kace — APPROVED** [review]

**Review by kace — APPROVED** [review]

**Comment by arjkesh** [comment]
Closing this in favor of https://github.com/aws/deep-learning-containers/pull/2891/files
